src/bootstrap/format.rs

```diff
@@ -1,7 +1,196 @@
 //! Runs rustfmt on the repository.
 
-use crate::Build;
+use {
+    crate::{util, Build, Crate},
+    glob::glob,
+    serde::Deserialize,
+    std::{
+        collections::BTreeSet,
+        fs,
+        path::{Path, PathBuf},
+        process::Command,
+    },
+};
 
 pub fn format(build: &Build, check: bool) {
-    unimplemented!("do the format command... {:?} {:?}", build.src, check);
+    let target = std::env::var("BUILD").expect("BUILD var is set by bootstrap.py");
+    let rustfmt_path = build.initial_rustc.with_file_name(util::exe("rustfmt", &target));
+    assert!(rustfmt_path.is_file(), "{} not a file", rustfmt_path.display());
+
+    let ignores = RustfmtIgnore::read(build);
+    let mut cmd = Command::new(rustfmt_path);
+    cmd.current_dir(&build.src);
+    ignores.add_args(&mut cmd);
+
+    if check {
+        cmd.arg("--check");
+    }
+
+    // the rest of the arguments are just the file entrypoints into the tree
+    for krate in build.crates.values().filter(|c| c.is_local(build)) {
+        for path in crate_entrypoint_paths(build, krate) {
+            if !ignores.is_ignored(&path) {
+                cmd.arg(path);
+            }
+        }
+    }
+
+    let status = cmd.status().expect("executing rustfmt");
+    assert!(status.success(), "rustfmt errored with status {:?}", status);
+}
+
+/// Gets all of the files that rustfmt needs to receive as arguments to format a crate.
+///
+/// Searches for files based on [Cargo's behavior]. This means looking in `Cargo.toml` for the
+/// `[bin]`, `[lib]`, `[bench]`, `[test]`, and `[example]` top-level keys and searching in the
+/// supported auto-discovered locations:
+///
+/// ```text
+/// ▾ src/           # directory containing source files
+///   lib.rs         # the main entry point for libraries and packages
+///   main.rs        # the main entry point for packages producing executables
+///   ▾ bin/         # (optional) directory containing additional executables
+///     *.rs
+///   ▾ */           # (optional) directories containing multi-file executables
+///     main.rs
+/// ▾ examples/      # (optional) examples
+///   *.rs
+///   ▾ */           # (optional) directories containing multi-file examples
+///     main.rs
+/// ▾ tests/         # (optional) integration tests
+///   *.rs
+///   ▾ */           # (optional) directories containing multi-file tests
+///     main.rs
+/// ▾ benches/       # (optional) benchmarks
+///   *.rs
+///   ▾ */           # (optional) directories containing multi-file benchmarks
+///     main.rs
+/// ```
+///
+/// [Cargo's behavior]: https://doc.rust-lang.org/cargo/reference/manifest.html#configuring-a-target
+fn crate_entrypoint_paths(build: &Build, krate: &Crate) -> BTreeSet<PathBuf> {
```

Comment:
Is there a reason we can't use `cargo fmt` here instead of rolling our own logic?

Reply:
I wasn't sure how to download it with the stage0 helper code -- tips?

Reply:
Isn't `cargo-fmt` part of the `rustfmt` component?

Reply:
So it is!